src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -550,6 +579,61 @@ Node *RegionNode::Ideal(PhaseGVN *phase, bool can_reshape) {
           }
         }
       }
+    } else {
+      // in passive materialization, we create a control node which covers multiple inputs of a region node.
+      //
+      // +-----------+   +--------+
+      // | CFG node  |   |cfg n   |
+      // +-----------+   +--------+
+      //      |  \  \     /
+      //      |   |  |   |
+      // +----------------+
+      // | Region         |
+      // +----------------+
+      //
+      // we need to normalize the region node, and its phi nodes.
```

Comment:
let's say CFG Node = 100. CFG n = 101 and region is 200. 
because of passive materialization, we have a region node like this.  CFG Node 100 covers 3 paths of 200 Region.  those 3 paths share the same 'materialization code snippet'.
 
```
200 Region === [200, 100, 100, 100, 101]
```

we expect to see.
```
200 Region === [200, 100, 101]
```

This is known as 'normalization'.  200 Region === [200, 100, 100, 100, 101] isn't valid form for build_cfg. C2 codegen will run into problem. that's why we have to normalize in optimizer.

Reply:
Not duplicated inputs. essentially, we hoist a common factor. 
Each input represents one predecessor of the basic block. we materialize the object once for all of them.